### DISCUSSION TITLE
Landis+Gyr E450 Encrypted smart meter data (STW Klagenfurt)

### DISCUSSION THREAD

**Discussion by andreas-doppelhofer** [arendst/Tasmota#21359] [Q&A]
Hi all, i got my smart meter a few days ago and already requested the "Kundenschnittstelle" from STW Klagenfurt. 
The interface is now activated and the wired M-Bus interface running using M-BUS SLAVE CLICK and NODEMCU ESP32. 
I compiled the firmware and applied it on the device and tried to setup everything to get data out.

Based on the description of STW the data (EN 13757-1: 2005 dlms/COSEM) should be encrypted (i have the key and decryption works, at least i have a output in the debug log)

I'm using following script to retrieve data, but the data is not correct. I get values but those are somehow not correct.

Tasmota script:

>D 48
>B
print Before senors
=>sensor53 r
tper=5
>M 1
+1,16,r,0,2400,Haus
1,=so3,512
1,=so4,MY-KEY
1,=s06,5000
; Test some starting positions
1,020f01x309UUuuUUuu@1,Test,x,Test,0
1,020f01x310UUuuUUuu@1,Test,x,Test,0
1,020f01x311UUuuUUuu@1,Test,x,Test,0
1,020f01x312UUuuUUuu@1,Test,x,Test,0
1,020f01x313UUuuUUuu@1,Test,x,Test,0
1,020f01x314UUuuUUuu@1,Test,x,Test,0
#
#

![image](https://github.com/arendst/Tasmota/assets/43139367/268aa87d-a5d3-4eb2-a5a5-54b87fe8b0bd)

Output of _sensor53 d1_
`14:25:42.236 SML: decrypted block: 355 bytes`
```
# formatted by a script of mine... #
02 0f 01 0f 02 04 12 00 28 09 06 00 08 19 09 00 ff
0f 02 12 00 00 02 04 12 00 28 09 06 00 08 19 09 00 ff
0f 01 12 00 00 02 04 12 00 01 09 06 00 00 60 01 00 ff
0f 02 12 00 00 02 04 12 00 08 09 06 00 00 01 00 00 ff
0f 02 12 00 00 02 04 12 00 03 09 06 01 00 01 07 00 ff
0f 02 12 00 00 02 04 12 00 03 09 06 01 00 02 07 00 ff
0f 02 12 00 00 02 04 12 00 03 09 06 01 00 01 08 00 ff
0f 02 12 00 00 02 04 12 00 03 09 06 01 00 02 08 00 ff
0f 02 12 00 00 02 04 12 00 03 09 06 01 00 03 07 00 ff
0f 02 12 00 00 02 04 12 00 03 09 06 01 00 04 07 00 ff
0f 02 12 00 00 02 04 12 00 03 09 06 01 00 03 08 00 ff
0f 02 12 00 00 02 04 12 00 03 09 06 01 00 04 08 00 ff
0f 02 12 00 00 02 04 12 00 03 09 06 01 00 1f 07 00 ff
0f 02 12 00 00 02 04 12 00 03 09 06 01 00 33 07 00 ff
0f 02 12 00 00 02 04 12 00 03 09 06 01 00 47 07 00 ff
0f 02 12 00 00 09 06 00 08 19 09 00 ff
09 08 36 38 33 33 38 33 30 31 09 0c 07 e8 05 05 07 13 2b 0a ff
80 00 80 06 00 00 01 42 06 00 00 00 00 06 00 06 96 5a 06 00 00 00 00 06 00 00 00 1c 06 00 00 00 40 06 00 03 d6 3e 06 00 00 08 7c 12 00 2a 12 00 55 12 00 2e 12 00 2e
```

I can find all available OBIS codes in the output (like 1.7.0, 1.8.0...) at the beginning (by converting from hex to DEC) but the data, most likely in the last 2 lines doesn't make sense. All other hex values above stays the same compared to the day before.
 I was assuming _06_ is the delimiter in the last 2 lines. But looks like this is not correct.

Any idea how to configure the filters in a correct way?

Thx Andi

**Comment by andreas-doppelhofer** [comment]
@gemu2015 as i saw a lot of request from your side in this bug https://github.com/arendst/Tasmota/issues/8790

Do you have an idea?

**Reply by gemu2015** [reply]
did you check this example in the sml docs ?
2 * Landis+Gyr E450 (encrypted)
seems to be very similar to yours

since i do not have any encrypted meter i don't know what the issue might be

**Reply by andreas-doppelhofer** [reply]
Yes, that's why i started with Tasmota. I red "encrypted" and hey this might work, but sadly it doesn't. Only **0** values in the result.
And my example above is using the _1,020f01x309UUuuUUuu@1,Test,x,Test,0_ syntax because with that approach i even can get values out but sadly not the correct ones, maybe from the wrong position or whatever. Yesterday i played around starting with ignoring only 1,2,3,4.... bytes to retrieve the values but at some position somewhere in the middle it starts collecting garbage values.

**Reply by cxcsimon** [reply]
Hey, I got the same problem here, I also have got the Landis Gyr E450 from Stadtwerke Klagenfurt. I got a simple M Bus Slave to USB converter and I get the still engrypted data as an output (starting with 7E A088 and also ending with 7E, so the data should be correct). I just have no clue how to decrypt the data....

What program do you use for decryption?

Best greets,
Simon

**Reply by andreas-doppelhofer** [reply]
I'm decrypting the message with this command in my script (see above). But as said i do not get correct values...
`1,=so4,MY-KEY`

**Reply by andreas-doppelhofer** [reply]
This is some more debug log from the console... I don't know how if this is ok or not.
```
17:25:41.785 BRY: GC from 4827 to 4231 bytes, objects freed 2/46 (in 1 ms) - slots from 51/122 to 51/122
17:25:41.791 RSL: STATE = {"Time":"2024-05-14T17:25:41","Uptime":"0T00:27:16","UptimeSec":1636,"Heap":132,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":410,"MqttCount":0,"Berry":{"HeapUsed":4,"Objects":46},"Wifi":{"AP":1,"SSId":"dops.at","BSSId":"B4:B0:24:4B:05:D8","Channel":2,"Mode":"11n","RSSI":92,"Signal":-54,"LinkCount":1,"Downtime":"0T00:00:04"}}
17:25:41.838 RSL: SENSOR = {"Time":"2024-05-14T17:25:41","Haus":{"Test":18,"Test":0,"Test":0,"Test":2,"Test":4,"Test":18,"Test":1,"Test":0,"Test":0,"Test":255,"Test":15,"Test":7,"Test":7,"Test":7,"Test":0,"Test":255,"Test":15,"Test":2,"Test":18,"Test":0,"Test":0,"Test":2}}
17:25:44.881 SML: HDLC frame:
17:25:44.883 SML: LLC frame:
17:25:44.884 SML: GBT frame:
17:25:45.578 SML: HDLC frame:
17:25:45.579 SML: GBT frame:
17:25:46.275 SML: HDLC frame:
17:25:46.276 SML: GBT frame:
17:25:46.619 SML: HDLC frame:
17:25:46.620 SML: GBT frame:
17:25:46.624 SML: GCM frame:
17:25:46.625 SML: DLMS frame:
17:25:46.627 SML: Received valid DLMS at 45
17:25:46.628 SML: Using application data:
17:25:46.630 SML: decrypted block: 355 bytes
17:25:46.632 :>02 0f 01 0f 02 04 12 00 28 09 06 00 08 19 09 00 
17:25:46.644 :>ff 0f 02 12 00 00 02 04 12 00 28 09 06 00 08 19 
17:25:46.646 :>09 00 ff 0f 01 12 00 00 02 04 12 00 01 09 06 00 
17:25:46.658 :>00 60 01 00 ff 0f 02 12 00 00 02 04 12 00 08 09 
17:25:46.660 :>06 00 00 01 00 00 ff 0f 02 12 00 00 02 04 12 00 
17:25:46.662 :>03 09 06 01 00 01 07 00 ff 0f 02 12 00 00 02 04 
17:25:46.674 :>12 00 03 09 06 01 00 02 07 00 ff 0f 02 12 00 00 
17:25:46.676 :>02 04 12 00 03 09 06 01 00 01 08 00 ff 0f 02 12 
17:25:46.678 :>00 00 02 04 12 00 03 09 06 01 00 02 08 00 ff 0f 
17:25:46.690 :>02 12 00 00 02 04 12 00 03 09 06 01 00 03 07 00 
17:25:46.692 :>ff 0f 02 12 00 00 02 04 12 00 03 09 06 01 00 04 
17:25:46.704 :>07 00 ff 0f 02 12 00 00 02 04 12 00 03 09 06 01 
17:25:46.706 :>00 03 08 00 ff 0f 02 12 00 00 02 04 12 00 03 09 
17:25:46.708 :>06 01 00 04 08 00 ff 0f 02 12 00 00 02 04 12 00 
17:25:46.720 :>03 09 06 01 00 1f 07 00 ff 0f 02 12 00 00 02 04 
17:25:46.723 :>12 00 03 09 06 01 00 33 07 00 ff 0f 02 12 00 00 
17:25:46.726 :>02 04 12 00 03 09 06 01 00 47 07 00 ff 0f 02 12 
17:25:46.739 :>00 00 09 06 00 08 19 09 00 ff 09 08 36 38 33 33 
17:25:46.741 :>38 33 30 31 09 0c 07 e8 05 0e 02 12 19 28 ff 80 
17:25:46.744 :>00 80 06 00 00 03 60 06 00 00 00 00 06 00 07 6e 
17:25:46.756 :>ad 06 00 00 00 00 06 00 00 00 00 06 00 00 00 c3 
17:25:46.758 :>06 00 03 d8 30 06 00 00 2c a4 12 01 32 12 00 49 
17:25:46.760 :>12 00 20 
17:25:46.761 :>12 00 20 
17:25:51.428 :>02 04 12 00 03 09 06 01 00 01 08 00 ff 0f 02 12 
17:25:51.430 :>00 00 02 04 12 00 03 09 06 01 00 02 08 00 ff 0f 
17:25:51.442 :>02 12 00 00 02 04 12 00 03 09 06 01 00 03 07 00 
17:25:51.444 :>ff 0f 02 12 00 00 02 04 12 00 03 09 06 01 00 04 
17:25:51.456 :>07 00 ff 0f 02 12 00 00 02 04 12 00 03 09 06 01 
17:25:51.458 :>00 03 08 00 ff 0f 02 12 00 00 02 04 12 00 03 09 
17:25:51.460 :>06 01 00 04 08 00 ff 0f 02 12 00 00 02 04 12 00 
17:25:51.472 :>03 09 06 01 00 1f 07 00 ff 0f 02 12 00 00 02 04 
17:25:51.475 :>12 00 03 09 06 01 00 33 07 00 ff 0f 02 12 00 00 
17:25:51.478 :>02 04 12 00 03 09 06 01 00 47 07 00 ff 0f 02 12 
17:25:51.491 :>00 00 09 06 00 08 19 09 00 ff 09 08 36 38 33 33 
17:25:51.493 :>38 33 30 31 09 0c 07 e8 05 0e 02 12 19 2d ff 80 
17:25:51.496 :>00 80 06 00 00 03 60 06 00 00 00 00 06 00 07 6e 
17:25:51.508 :>ae 06 00 00 00 00 06 00 00 00 00 06 00 00 00 c3 
17:25:51.510 :>06 00 03 d8 30 06 00 00 2c a4 12 01 32 12 00 49 
17:25:51.512 :>12 00 20 
17:25:51.513 :>12 00 20 
17:25:54.404 SML: HDLC frame:
17:25:54.406 SML: LLC frame:
17:25:54.407 SML: GBT frame:
17:25:55.101 SML: HDLC frame:
17:25:55.102 SML: GBT frame:
17:25:55.798 SML: HDLC frame:
17:25:55.799 SML: GBT frame:
17:25:56.142 SML: HDLC frame:
17:25:56.143 SML: GBT frame:
17:25:56.147 SML: GCM frame:
17:25:56.148 SML: DLMS frame:
17:25:56.150 SML: Received valid DLMS at 45
17:25:56.151 SML: Using application data:
17:25:56.153 SML: decrypted block: 355 bytes
17:25:56.155 :>02 0f 01 0f 02 04 12 00 28 09 06 00 08 19 09 00 
17:25:56.167 :>ff 0f 02 12 00 00 02 04 12 00 28 09 06 00 08 19 
17:25:56.169 :>09 00 ff 0f 01 12 00 00 02 04 12 00 01 09 06 00 
17:25:56.181 :>00 60 01 00 ff 0f 02 12 00 00 02 04 12 00 08 09 
17:25:56.183 :>06 00 00 01 00 00 ff 0f 02 12 00 00 02 04 12 00 
17:25:56.185 :>03 09 06 01 00 01 07 00 ff 0f 02 12 00 00 02 04 
17:25:56.197 :>12 00 03 09 06 01 00 02 07 00 ff 0f 02 12 00 00 
17:25:56.199 :>02 04 12 00 03 09 06 01 00 01 08 00 ff 0f 02 12 
17:25:56.201 :>00 00 02 04 12 00 03 09 06 01 00 02 08 00 ff 0f 
17:25:56.213 :>02 12 00 00 02 04 12 00 03 09 06 01 00 03 07 00 
17:25:56.215 :>ff 0f 02 12 00 00 02 04 12 00 03 09 06 01 00 04 
17:25:56.227 :>07 00 ff 0f 02 12 00 00 02 04 12 00 03 09 06 01 
17:25:56.229 :>00 03 08 00 ff 0f 02 12 00 00 02 04 12 00 03 09 
17:25:56.231 :>06 01 00 04 08 00 ff 0f 02 12 00 00 02 04 12 00 
17:25:56.243 :>03 09 06 01 00 1f 07 00 ff 0f 02 12 00 00 02 04 
17:25:56.245 :>12 00 03 09 06 01 00 33 07 00 ff 0f 02 12 00 00 
17:25:56.247 :>02 04 12 00 03 09 06 01 00 47 07 00 ff 0f 02 12 
17:25:56.258 :>00 00 09 06 00 08 19 09 00 ff 09 08 36 38 33 33 
17:25:56.259 :>38 33 30 31 09 0c 07 e8 05 0e 02 12 19 32 ff 80 
17:25:56.271 :>00 80 06 00 00 03 5f 06 00 00 00 00 06 00 07 6e 
17:25:56.272 :>af 06 00 00 00 00 06 00 00 00 00 06 00 00 00 c3 
17:25:56.284 :>06 00 03 d8 30 06 00 00 2c a5 12 01 32 12 00 49 
17:25:56.285 :>12 00 20 
17:25:56.286 :>12 00 20 
17:25:57.327 WIF: Checking connection...
17:25:57.334 BRY: GC from 4827 to 4231 bytes, objects freed 2/46 (in 1 ms) - slots from 51/122 to 51/122
17:25:57.340 RSL: STATE = {"Time":"2024-05-14T17:25:57","Uptime":"0T00:27:32","UptimeSec":1652,"Heap":132,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":248,"MqttCount":0,"Berry":{"HeapUsed":4,"Objects":46},"Wifi":{"AP":1,"SSId":"dops.at","BSSId":"B4:B0:24:4B:05:D8","Channel":2,"Mode":"11n","RSSI":92,"Signal":-54,"LinkCount":1,"Downtime":"0T00:00:04"}}
17:25:57.387 RSL: SENSOR = {"Time":"2024-05-14T17:25:57","Haus":{"Test":18,"Test":0,"Test":0,"Test":2,"Test":4,"Test":18,"Test":1,"Test":0,"Test":0,"Test":255,"Test":15,"Test":7,"Test":7,"Test":7,"Test":0,"Test":255,"Test":15,"Test":2,"Test":18,"Test":0,"Test":0,"Test":2}}
17:25:59.656 SML: HDLC frame:
17:25:59.658 SML: LLC frame:
17:25:59.659 SML: GBT frame:
17:26:00.353 SML: HDLC frame:
17:26:00.354 SML: GBT frame:
17:26:01.050 SML: HDLC frame:
17:26:01.052 SML: GBT frame:
17:26:01.394 SML: HDLC frame:
17:26:01.396 SML: GBT frame:
17:26:01.399 SML: GCM frame:
17:26:01.401 SML: DLMS frame:
17:26:01.402 SML: Received valid DLMS at 45
17:26:01.404 SML: Using application data:
17:26:01.406 SML: decrypted block: 355 bytes
17:26:01.407 :>02 0f 01 0f 02 04 12 00 28 09 06 00 08 19 09 00 
17:26:01.419 :>ff 0f 02 12 00 00 02 04 12 00 28 09 06 00 08 19 
17:26:01.421 :>09 00 ff 0f 01 12 00 00 02 04 12 00 01 09 06 00 
17:26:01.433 :>00 60 01 00 ff 0f 02 12 00 00 02 04 12 00 08 09 
17:26:01.435 :>06 00 00 01 00 00 ff 0f 02 12 00 00 02 04 12 00 
17:26:01.437 :>03 09 06 01 00 01 07 00 ff 0f 02 12 00 00 02 04 
17:26:01.449 :>12 00 03 09 06 01 00 02 07 00 ff 0f 02 12 00 00 
17:26:01.451 :>02 04 12 00 03 09 06 01 00 01 08 00 ff 0f 02 12 
17:26:01.453 :>00 00 02 04 12 00 03 09 06 01 00 02 08 00 ff 0f 
17:26:01.465 :>02 12 00 00 02 04 12 00 03 09 06 01 00 03 07 00 
17:26:01.467 :>ff 0f 02 12 00 00 02 04 12 00 03 09 06 01 00 04 
17:26:01.479 :>07 00 ff 0f 02 12 00 00 02 04 12 00 03 09 06 01 
17:26:01.481 :>00 03 08 00 ff 0f 02 12 00 00 02 04 12 00 03 09 
17:26:01.484 :>06 01 00 04 08 00 ff 0f 02 12 00 00 02 04 12 00 
17:26:01.497 :>03 09 06 01 00 1f 07 00 ff 0f 02 12 00 00 02 04 
17:26:01.499 :>12 00 03 09 06 01 00 33 07 00 ff 0f 02 12 00 00 
17:26:01.502 :>02 04 12 00 03 09 06 01 00 47 07 00 ff 0f 02 12 
17:26:01.514 :>00 00 09 06 00 08 19 09 00 ff 09 08 36 38 33 33 
17:26:01.517 :>38 33 30 31 09 0c 07 e8 05 0e 02 12 19 37 ff 80 
17:26:01.519 :>00 80 06 00 00 03 60 06 00 00 00 00 06 00 07 6e 
17:26:01.532 :>b1 06 00 00 00 00 06 00 00 00 00 06 00 00 00 c3 
17:26:01.533 :>06 00 03 d8 30 06 00 00 2c a5 12 01 32 12 00 49 
17:26:01.535 :>12 00 20 
17:26:01.537 :>12 00 20 
```

**Reply by andreas-doppelhofer** [reply]
What i know from calculating these hex values into dec, my "Zählerstand" in kWh is located at that position. But no matter how i use the offset i'm always wrong...
![image](https://github.com/arendst/Tasmota/assets/43139367/5afbe286-941f-438c-bbd3-a0f164251c51)

**Reply by fkroepfl** [reply]
I don't know if this is helpful, but it seems that there were/are problems with "Energie Klagenfurt" meters here too:

https://forum.shrdzm.com/viewtopic.php?t=245

The same applies to the P1 meter from https://www.homewizard.com/de-at/, as I was informed.

**Reply by fkroepfl** [reply]
What I have heard, but unfortunately cannot verify, is that the STW Klagenfurt meter first sends 15 lines with the Obis codes and then 15 lines with the corresponding values.
 
Meters of other operators, if I have understood correctly, send Obis code and the value in the same line.

Under certain circumstances, this is one of the reasons why the datagrams cannot be decoded so easily.

**Reply by fkroepfl** [reply]
Under certain circumstances, this service can also help a little and provide additional information, or?

- https://tasmota-sml-parser.dicp.net/

**Reply by andreas-doppelhofer** [reply]
First success after some debugging! :sunglasses: 
![image](https://github.com/arendst/Tasmota/assets/43139367/3c1f44a2-ffd6-41c5-a0db-ef9aec2925ff)

It is not possible to set an offset > 255 -> [xsns_53_sml.ino](https://github.com/arendst/Tasmota/blob/development/tasmota/tasmota_xsns_sensor/xsns_53_sml.ino#L2025)
By changing this to uint16_t :partying_face:  FYI @gemu2015 
```
@@ -2007,8 +2024,10 @@ void SML_Decode(uint8_t index) {
               } else {
                 mp++;
                 if (isdigit(*mp)) {
-                  uint8_t skip = strtol((char*)mp, (char**)&mp, 10);
+                  uint16_t skip = strtol((char*)mp, (char**)&mp, 10);
                   cp += skip;

```
Now i need to find time to get everything running....

**Reply by andreas-doppelhofer** [reply]
Sadly i just have 2 values in my stream which is valuable. 

- Current 
- Power consumption in total

Nothing more... Need to check with STW.

**Reply by fkroepfl** [reply]
What do the specialists at Stadtwerke Klagenfurt have to say?

**Reply by andreas-doppelhofer** [reply]
Nothing, they do not really know how it should look like. They had configured the protocol a few years ago with Landis but do not really have a clue whats going on. I asked them for a protocol description but i haven't received anything yet.

**Comment by fkroepfl** [comment]
> The interface is now activated and the wired M-Bus interface running using M-BUS SLAVE CLICK and NODEMCU ESP32.

The Mbus interface has now also been activated on our E450 from STW/Energie Klagenfurt. I don't have a hardware adapter yet and am looking for the right components to work with the Tasmota Smart Meter feature.

Can you share your source of supply, interconnection and configuration of the components here?

**Reply by andreas-doppelhofer** [reply]
I will do if i have a working one...

**Comment by fkroepfl** [comment]
I can currently receive the following data telegrams from the „Kundenschnittstelle (Mbus)" of our "STW Klagenfurt" meter. (see screenshot)

@andreas-doppelhofer: 

- Do the data telegrams correspond to the ones you are receiving?
- Do you already know what the data telegram structure of the "STW Klagenfurt" looks like?

<img width="1432" alt="Bildschirmfoto 2024-05-28 um 18 00 44" src="https://github.com/arendst/Tasmota/assets/6333880/e09a2c2b-abe4-45e4-8b4b-e122b95ee31f">

**Reply by fkroepfl** [reply]
It looks like I have found the OBIS registers in the data telegram. Nevertheless, it is not clear to me where I can find the corresponding values.

Any idea?


![image](https://github.com/arendst/Tasmota/assets/6333880/dd2d9f37-1868-4f20-b9da-0f0bf6c796ed)

<img width="711" alt="image" src="https://github.com/arendst/Tasmota/assets/6333880/0e16689b-7c33-4f6a-9150-fca56dc30823">

**Reply by fkroepfl** [reply]
Taking into account all the information available to me and understood by me, I come to the following result when manually interpreting the above-mentioned data set.

Can this make sense?

![image](https://github.com/arendst/Tasmota/assets/6333880/6cef3f50-bf3a-47c0-9d33-e385eea159aa)

**Reply by andreas-doppelhofer** [reply]
Yes those values look like the same to my power meter.

**Reply by andreas-doppelhofer** [reply]
This is my current working script. But only with power consumption and current (maybe L1)
(To get it running, do not forget to change uint8_t -> uint16_t see above)
```
>D 48
pin=0
pusage=0
plast=0
mystamp=0
>B
print Before senors
=>sensor53 r
tper=5
>T
pin=Haus#Total
>S
mystamp=tsn(tstamp)
mystamp-=3600
if plast==0 {
  pusage=0
  plast=pin
}

if upsecs%600==0 {
    pusage=pin-plast
    plast=pin
    print %upsecs% %pin% %pusage% %plast% %s(2.0mystamp)%
}
>M 1
+1,16,r,0,2400,Haus
1,=so3,512
1,=so4,xxxxxxxxxxxxxxxxxxxxxxx
1,=s06,5000
1,020f01x314UUuuUUuu@1,wh,wh,Total,3
1,020f01x304UUuuUUuu@1000,CurrentIn,A,CurrentIn,3
1,020f01x309UUuuUUuu@1000,value2,kwh,value2,3
1,020f01x334UUuuUUuu@1000,value3,kwh,value3,3
#
```

**Reply by andreas-doppelhofer** [reply]
I'm using WebQuery/Send in the script above (removed the code for simplicity) and push it to my database. The power consumption will be pushed every 5min.. see my chart
![image](https://github.com/arendst/Tasmota/assets/43139367/ce19074f-bb85-4439-8a28-98edf13c676c)

**Reply by fkroepfl** [reply]
> 1,020f01x314UUuuUUuu@1,wh,wh,Total,3
> 1,020f01x304UUuuUUuu@1000,CurrentIn,A,CurrentIn,3
> 1,020f01x309UUuuUUuu@1000,value2,kwh,value2,3
> 1,020f01x334UUuuUUuu@1000,value3,kwh,value3,3

Many thanks for the information. 

One question: If I compare this with my data telegram, I get different offsets. Do you have any idea where the difference could come from? Am I assuming a wrong sequence of the OBIS values?

![image](https://github.com/arendst/Tasmota/assets/6333880/9094c482-2808-49f1-b01e-910c8664284d)

**Reply by fkroepfl** [reply]
> (To get it running, do not forget to change uint8_t -> uint16_t see above)

@andreas-doppelhofer Could you provide me with a modified binary file for ESP32 and ESP32V3?
Unfortunately, I have no experience of how to do this. Sorry.

**Reply by fkroepfl** [reply]
> 1,=s06,5000

@andreas-doppelhofer Is this a typo or is it intentional?
I can only find the entry `M,=so6` at https://tasmota.github.io/docs/Smart-Meter-Interface/.

**Reply by andreas-doppelhofer** [reply]
No, this is no type. You need to set the Meter ID, in that case 1.
```
+1,16,r,0,2400,Haus <---------+1 meter number one
1,=s06              <--------- set options for meter number one
```

**Reply by andreas-doppelhofer** [reply]
> One question: If I compare this with my data telegram, I get different offsets. Do you have any idea where the difference could come from? Am I assuming a wrong sequence of the OBIS values?

Can you provide the decrypted data in plaintext (like the one in your first screenshot)? Then it's easier to feed my script for checking the offset.

**Reply by andreas-doppelhofer** [reply]
> @andreas-doppelhofer Could you provide me with a modified binary file for ESP32 and ESP32V3?
Unfortunately, I have no experience of how to do this. Sorry.

This firmware is compiled for ESP32_DevKitc (in my case, i'm using v4)
I already set dummy Wifi credentials: (change to your Wifi, or use this to just check if everything is working, no hardcoded IP is set, DHCP need to do the magic)
* SSID: **IOT**
* Password: **!AbC#1234567890**
[tasmota32.factory.bin.gz](https://github.com/user-attachments/files/15510760/tasmota32.factory.bin.gz)

**Reply by fkroepfl** [reply]
> No, this is no type. You need to set the Meter ID, in that case 1.

Ok, thx. But there's nothing about that in the Tasmota online documentation, is there?

**Reply by dops-at** [reply]
Had a look at the examples... It is also mentioned slightly in the docs
![image](https://github.com/arendst/Tasmota/assets/16140631/1f8efba6-a996-4e04-871b-7df860c544ca)

**Reply by andreas-doppelhofer** [reply]
Just for reference, I opened a PR for setting uint16_t as default. see https://github.com/arendst/Tasmota/pull/21535

**Reply by fkroepfl** [reply]
> Can you provide the decrypted data in plaintext (like the one in your first screenshot)? Then it's easier to feed my script for checking the offset.

Does this work for you?

<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:////Users/cipo/Library/Group%20Containers/UBF8T346G9.Office/TemporaryItems/msohtmlclip/clip.htm">
<link rel=File-List
href="file:////Users/cipo/Library/Group%20Containers/UBF8T346G9.Office/TemporaryItems/msohtmlclip/clip_filelist.xml">
<style>

</style>
</head>

<body link="#0563C1" vlink="#954F72">


SML:   decrypted block: 355 bytes | OBIS Register;Index | description | unit | alias
-- | -- | -- | -- | --
020f010f |   |   |   |  
020412002809060008190900ff0f02120000 | 0-8:25.9.0;2 |   |   |  
020412002809060008190900ff0f01120000 | 0-8:25.9.0;1 |   |   |  
020412000109060000600100ff0f02120000 | 0-0:96.1.0;2 |   |   |  
020412000809060000010000ff0f02120000 | 0-0:1.0.0;2 |   |   |  
020412000309060100010700ff0f02120000 | 1-0:1.7.0;2 |   |   |  
020412000309060100020700ff0f02120000 | 1-0:2.7.0;2 |   |   |  
020412000309060100010800ff0f02120000 | 1-0:1.8.0;2 |   |   |  
020412000309060100020800ff0f02120000 | 1-0:2.8.0;2 |   |   |  
020412000309060100030700ff0f02120000 | 1-0:3.7.0;2 |   |   |  
020412000309060100040700ff0f02120000 | 1-0:4.7.0;2 |   |   |  
020412000309060100030800ff0f02120000 | 1-0:3.8.0;2 |   |   |  
020412000309060100040800ff0f02120000 | 1-0:4.8.0;2 |   |   |  
0204120003090601001f0700ff0f02120000 | 1-0:31.7.0;2 |   |   |  
020412000309060100330700ff0f02120000 | 1-0:51.7.0;2 |   |   |  
020412000309060100470700ff0f02120000 | 1-0:71.7.0;2 |   |   |  
09060008190900ff | 0-8:25.9.0 | List of pushed objects |   |  
09083632383534323235 | 0-0:96.1.0 | Meter serial number |   |  
090c07e8051c020e0523ff800080 | 0-0:1.0.0 | Meter date   & time |   |  
060000074a | 1-0:1.7.0 | Positive active instantaneous power | kW | P+
0600000000 | 1-0:2.7.0 | Negative active instantaneous power | kW | P-
0600bb1951 | 1-0:1.8.0 | Positive active energy, total | kWh | A+
06001bf3db | 1-0:2.8.0 | Negative active energy, total | kWh | A-
0600000202 | 1-0:3.7.0 | Positive reactive instantaneous power | kW | P+
0600000000 | 1-0:4.7.0 | Negative reactive instantaneous power | kW | P-
06004cea5a | 1-0:3.8.0 | Positive reactive energy, total | kvarh | Q+
06000fb2ab | 1-0:4.8.0 | Negative reactive energy, total | kvarh | Q-
1200ae | 1-0:31.7.0 | Instantaneous current in phase I | A | I (L1)
1200c2 | 1-0:51.7.0 | Instantaneous current in phase II | A | I (L2)
1201f5 | 1-0:71.7.0 | Instantaneous current in phase III | A | I (L3)
1201f5 |   |   |   |  



</body>

</html>

**Reply by dops-at** [reply]
Just this as plain text, would be fine...
![image](https://github.com/arendst/Tasmota/assets/16140631/0c434aca-0965-425c-a1b9-5c4ce6ebe617)

**Reply by fkroepfl** [reply]
Plain enough? 😉

```
020f010f
020412002809060008190900ff0f02120000
020412002809060008190900ff0f01120000
020412000109060000600100ff0f02120000
020412000809060000010000ff0f02120000
020412000309060100010700ff0f02120000
020412000309060100020700ff0f02120000
020412000309060100010800ff0f02120000
020412000309060100020800ff0f02120000
020412000309060100030700ff0f02120000
020412000309060100040700ff0f02120000
020412000309060100030800ff0f02120000
020412000309060100040800ff0f02120000
0204120003090601001f0700ff0f02120000
020412000309060100330700ff0f02120000
020412000309060100470700ff0f02120000
09060008190900ff
09083632383534323235
090c07e8051c020e0523ff800080
060000074a
0600000000
0600bb1951
06001bf3db
0600000202
0600000000
06004cea5a
06000fb2ab
1200ae
1200c2
1201f5
1201f5

```

**Reply by andreas-doppelhofer** [reply]
As far as i can see the offset is the same as on my smart meter. So my script example above should work.

**Reply by cxcsimon** [reply]
I would have a working solution, see the message below

**Comment by Noschvie** [comment]
Did you have a look to [Tasmota SML decoder](https://tasmota-sml-parser.dicp.net/) ?

**Reply by fkroepfl** [reply]
Yes, but apparently this tool cannot do anything with this data telegram or decode parts of it.
Would have been too nice and easy 😉

**Comment by HofeBY** [comment]
@fkroepfl 
[https://github.com/HofeBY/Tasmota-Firmware](https://deref-gmx.net/mail/client/7XKPQCxIKjo/dereferrer/?redirectUrl=https%3A%2F%2Fgithub.com%2FHofeBY%2FTasmota-Firmware)

New Version ESP32 with STW-Klagenfurt patch uploded.
Also ESP32 RULE with MATTER .

Read Liesmich-zuerst.txt to build with new files for own builds .

HofeBY

**Reply by fkroepfl** [reply]
🙏 many thx, I guess it's this patch ->  https://github.com/arendst/Tasmota/commit/0249f66066e739fac2b4d7433be2cfb54ea1ec26 , or?

**Reply by HofeBY** [reply]
yes, included.
ESP32 is the right Prozessor type ?

**Reply by fkroepfl** [reply]
Yes, I'm using an ESP32.

**Reply by cxcsimon** [reply]
I would have a working solution, see the message below

**Comment by cxcsimon** [comment]
Hey, if you need an ESP32 based module that can encrypt and send energy data from Landis Gyr E450, you can contact me at: info@smarter-power.at

I have complete sets for 49,9€. I‘m currently still private, preparing for a commercial release.

and also sets with LoRa modules if you don‘t have WiFi at your energy meter. 

If you don‘t want to purchase a module I‘m also always happy to help :)

Kind regards,

Simon
